proposed/resource-limit.md

```diff
@@ -0,0 +1,843 @@
+
+However, we have not yet found any use cases for aggregated limiters in dotnet/runtime and hence it's not part of this proposal. If we eventually deem that such aggregated limiters are needed in the BCL as well, it can be added independently from the APIs proposed here.
+
+Sample API:
+
+```c#
+// Represent an aggregated resource (e.g. a resource limiter aggregated by IP)
+public abstract class AggregatedResourceLimiter<TKey>
+{
+    // an inaccurate view of resources
+    public abstract long EstimatedCount(TKey resourceID);
+
+    // Fast synchronous attempt to acquire resources
+    public abstract Resource Acquire(TKey resourceID, long requestedCount);
+
+    // Wait until the requested resources are available
+    // If unsuccessful, throw
+    public abstract ValueTask<Resource> AcquireAsync(TKey resourceID, long requestedCount, CancellationToken cancellationToken = default);
+}
+```
+
+See a [proof of concept implementation](#ip-address-aggregated-resource-limiter-using-a-local-concurrentdictionary-as-a-storage-mechanism).
+
+## Stakeholders and Reviewers
+
+- Owned by ASP.NET Core
+- Code will be a new area in dotnet/runtime
+  - Tentatively System.Threading.ResourceLimits
+- Reviewers
+  - David Fowler, Eric Erhardt, Stephen Halter, Chris Ross
+
+## Design
+
+### Abstractions
+
+The main abstraction API will consist of the `ResourceLimiter` and the associated `Resource`. The main responsibility of the `ResourceLimiter` is to allow the user to determine whether it's possible to proceed with an operation. The `Resource` struct is returned by the `ResourceLimiter` to represent if the acquisition was successful and the resources that were obtained, if successful.
+
+```c#
+    // Represents a limiter type that users interact with to determine if an operation can proceed
+    public abstract class ResourceLimiter
+    {
+        // An estimated count of the underlying resources
+        abstract long EstimatedCount { get; }
+
+        // Fast synchronous attempt to acquire resources
+        // Set requestedCount to 0 to get whether resource limit has been reached
+        abstract Resource Acquire(long requestedCount);
+
+        // Wait until the requested resources are available
+        // Set requestedCount to 0 to wait until resource is replenished
+        abstract ValueTask<Resource> AcquireAsync(long requestedCount, CancellationToken cancellationToken = default);
+    }
+
+    // Represents a resource obtained from the limiter. The user disposes this type to release the acquired resources.
+    public struct Resource : IDisposable
```

Comment:
Why is this defined twice?

Reply:
Are you referring to the section above ~line 150ish? That was an example of additional APIs required for partial release, but it's a non-goal for now. I've hidden it under <detail> tags for now.